modify the column type change doc (#5787)
pingcap/docs: sql-statements/sql-statement-modify-column.md

The `ALTER TABLE.. MODIFY COLUMN` statement modifies a column on an existing table. The modification can include changing the data type and attributes. To rename at the same time, use the [`CHANGE COLUMN`](/sql-statements/sql-statement-change-column.md) statement instead.

Since v5.1.0, TiDB has supported changes of data types for Reorg data, including but not limited to:

- Change from varchar to bigint
- Modify decimal precision

> - Length compression from varchar(10) to varchar(5)

Reviewer: ditto

1 row in set (0.00 sec)
```

### Reorg-Data Change

{{< copyable "sql" >}}

```sql
CREATE TABLE t1 (id int not null primary key AUTO_INCREMENT, col1 INT);
```

```
Query OK, 0 rows affected (0.11 sec)
```

{{< copyable "sql" >}}

```sql
INSERT INTO t1 (col1) VALUES (12345),(67890);
```

```
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0
```

{{< copyable "sql" >}}

```sql
ALTER TABLE t1 MODIFY col1 VARCHAR(5);
```

```
Query OK, 0 rows affected (2.52 sec)
```

{{< copyable "sql" >}}

```sql
SHOW CREATE TABLE t1\G;
```

```sql
*************************** 1. row ***************************
       Table: t1
CREATE TABLE `t1` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `col1` varchar(5) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=30001
1 row in set (0.00 sec)
```

> **Note:**
> 

Reviewer: ```suggestion
> **Note:**
>
```

> - TiDB returns an error when the changed data type conflicts with an existing data row. In the above example, TiDB returns the following error:
>

> >   ```
> >   alter table t1 modify column col1 varchar(4);
> >   ERROR 1406 (22001): Data Too Long, field len 4, data len 5
> >   ```

Reviewer: ```suggestion
>    ```
>    alter table t1 modify column col1 varchar(4);
>    ERROR 1406 (22001): Data Too Long, field len 4, data len 5
>    ```
```

>

> > - Due to the compatibility with the Async Commit feature, the DDL waits for a certain amount of time (about 2.5s) before starting to process into Reorg Data.

Reviewer: ```suggestion
> - Due to the compatibility with the Async Commit feature, the DDL statement waits for a period of time (about 2.5s) before starting to process into Reorg Data.
```

>

> >   ```
> >   Query OK, 0 rows affected (2.52 sec)
> >   ```

Reviewer: ```suggestion
>    ```
>    Query OK, 0 rows affected (2.52 sec)
>    ```
```

## MySQL compatibility

* Does not support modifying multiple columns using a single `ALTER TABLE` statement. For example:

    ```sql
    ALTER TABLE t1 MODIFY col1 BIGINT, MODIFY id BIGINT NOT NULL;
    ERROR 1105 (HY000): Unsupported multi schema change
    ```

> * Does not support modifying data types on the primary key columns. For example:

Reviewer: ```suggestion
* Does not support modifying the data types on the primary key columns. For example:
```

    ```sql
    CREATE TABLE t (a int primary key);
    ALTER TABLE t MODIFY COLUMN a VARCHAR(10);
    ERROR 8200 (HY000): Unsupported modify column: column has primary key flag
    ```

* Does not support modifying the types of generate columns. For example:

    ```sql

>     CREATE TABLE t1 (col1 BIGINT);
>     ALTER TABLE t1 MODIFY col1 INT;
>     ERROR 8200 (HY000): Unsupported modify column length 11 is less than origin 20

Reviewer: ```suggestion
```

    CREATE TABLE t (a INT, b INT as (a+1));
    ALTER TABLE t MODIFY COLUMN b VARCHAR(10);
    ERROR 8200 (HY000): Unsupported modify column: column is generated
    ```

* Does not support modifying the column types on the partitioned tables. For example:

    ```sql
    CREATE TABLE t (c1 INT, c2 INT, c3 INT) partition by range columns(c1) ( partition p0 values less than (10), partition p1 values less than (maxvalue));
    ALTER TABLE t MODIFY COLUMN c1 DATETIME;
    ERROR 8200 (HY000): Unsupported modify column: table is partition table
    ```

> * Does not support modifying some data types (for example, some TIME types, Bit, Set, Enum, JSON, etc.) are not supported due to some compatibility issues of TIDB cast function with MySQL's behavior.

Reviewer: ditto
